build_tools/cmake/iree_bytecode_module.cmake

```diff
@@ -114,6 +114,23 @@ function(iree_bytecode_module)
     list(APPEND _ARGS "-iree-llvm-embedded-linker-path=\"${_EMBEDDED_LINKER_TOOL_EXECUTABLE}\"")
   endif()
 
+  # Support testing in TSan build dirs. Unlike other sanitizers, TSan is an
+  # ABI break: when the host code is built with TSan, the module must be too,
+  # otherwise we get crashes calling module code.
+  #
+  # We make an exception for cross-compiling tests, crudely identified for now
+  # by the presence of a target-triple flag. IREE_ENABLE_TSAN implies (for the
+  # time being) a host build dir, so any such tests are by definition build-only
+  # which makes it unnecessary to enable TSan in module code. We could support
+  # that if needed, it's just that for now, as TSan linking is implemented using
+  # clang as the linker driver, we would have to figure out the right (-march?)
+  # flag to pass for cross-linking.
+  if (IREE_ENABLE_TSAN AND NOT (_ARGS MATCHES "-iree-llvm-target-triple="))
+    list(APPEND _ARGS "-iree-llvm-sanitize=thread")
```

Comment:
<sub>(adding as a comment on these lines so the conversation can be replied to directly and can be marked as "resolved", unlike top level PR comments)</sub>

> 2. In today's meeting we agreed that `IREE_ENABLE_TSAN` needs to be split into two settings, one enabling TSan in the compiler and the other enabling TSan in the runtime. Once that is done, the `iree_bytecode_module` logic debated above will just need to follow the runtime setting (which is where it's necessary, due to TSan breaking ABI) and that will thus become decoupled from the compiler. I will get back to this once https://reviews.llvm.org/D121182 is merged, and then, I will need to understand how exactly we implement that compiler/runtime decoupling --- based on source code directories or on individual cmake rules ?

How about keeping one set of options that passes `-fsanitize=[address,thread,memory]` to the CMake build system, affecting "runtime" binaries and "compiler" binaries: https://github.com/google/iree/blob/e9c786761f91f02e854875ca5b8b257cc706a993/build_tools/cmake/sanitizers.cmake#L16-L19

then adding a new set of options that adds `-iree-llvm-sanitize=[address,thread,memory]` to the CMake functions that invoke IREE's compiler.

Brainstorming names:
* `IREE_CODEGEN_ENABLE_TSAN` 
* `IREE_ENABLE_CODEGEN_TSAN`
* `IREE_TESTS_ENABLE_TSAN` 
* `IREE_ENABLE_TSAN_IN_TESTS`

Then sample flag usage:

Configuration | How to set flags
-- | --
Linux host + Linux target | Single CMake configure with both flags set
Linux host + Android target | Configure host with `IREE_ENABLE_TSAN_IN_TESTS`, target with `IREE_ENABLE_TSAN`

---

I don't think a split between "runtime binaries" and "compiler binaries" would make sense for this flag, like we have for the Tracing option, but here's how that is done:

https://github.com/google/iree/blob/e9c786761f91f02e854875ca5b8b257cc706a993/CMakeLists.txt#L35-L36

https://github.com/google/iree/blob/e9c786761f91f02e854875ca5b8b257cc706a993/iree/compiler/Utils/CMakeLists.txt#L58-L63

https://github.com/google/iree/blob/e9c786761f91f02e854875ca5b8b257cc706a993/iree/compiler/Utils/TracingUtils.h#L25-L30